server/controllers/finance/vouchers.js

```diff
@@ -0,0 +1,152 @@
+/**
+* The /vouchers/ HTTP API endpoint
+*
+* @module finance/vouchers
+*
+* @desc This module is responsible for handling all crud operations relatives
+* to voiuchers transactions and define all vouchers api functions
+*
+* @required node-uuid
+* @required lib/db
+*/
+
+'use strict';
+
+var _    = require('lodash'),
+    uuid = require('node-uuid'),
+    db   = require('../../lib/db');
+
+/** Get list of vouchers */
+exports.list = getList;
+
+/** Get detail of vouchers */
+exports.detail = getDetail;
+
+/** Create a new voucher record */
+exports.create = create;
+
+/**
+* GET /vouchers
+*
+* @method getList
+*/
+function getList(req, res, next) {
+  var query =
+    'SELECT v.uuid, v.date, v.project_id, v.reference, v.currency_id, v.amount, ' +
+    'v.description, v.document_uuid, v.user_id, vi.uuid AS voucher_item_uuid, ' +
+    'vi.account_id, vi.debit, vi.credit ' +
+    'FROM voucher v JOIN voucher_item vi ON vi.voucher_uuid = v.uuid ';
+
+  var builder = queryCondition(query, req.query);
+
+  db.exec(builder.query, builder.conditions)
+  .then(function (rows) {
+    res.status(200).json(rows);
+  })
+  .catch(next)
+  .done();
+}
+
+/**
+* GET /vouchers/:uuid
+*
+* @method getDetail
+*/
+function getDetail(req, res, next) {
+  var query =
+    'SELECT v.uuid, v.date, v.project_id, v.reference, v.currency_id, v.amount, ' +
+    'v.description, v.document_uuid, v.user_id, vi.uuid AS voucher_item_uuid, ' +
+    'vi.account_id, vi.debit, vi.credit ' +
+    'FROM voucher v JOIN voucher_item vi ON vi.voucher_uuid = v.uuid ' +
+    'WHERE v.uuid = ? ;';
+
+  db.exec(query, req.params.uuid)
+  .then(function (rows) {
+    if (!rows.length) { return next(new req.codes.ERR_NOT_FOUND()); }
+    res.status(200).json(rows[0]);
+  })
+  .catch(next)
+  .done();
+}
+
+/**
+* POST /vouchers
+*
+* @method create
+*/
+function create(req, res, next) {
+
+  if (!checkForMissing(req.body)) { return next(new req.codes.ERR_MISSING_REQUIRED_PARAMETERS()); }
+  if (req.body.voucher.date) { req.body.voucher.date = new Date(req.body.voucher.date); }
+
+  /** Initialise the transaction handler */
+  var transaction = db.transaction();
+
+  transaction.addQuery('INSERT INTO voucher SET ?', [req.body.voucher]);
+
+  var queryVoucherItem = _.isArray(req.body.voucher_item) ?
+    'INSERT INTO voucher_item (uuid, account_id, debit, credit, voucher_uuid) VALUES ?' :
+    'INSERT INTO voucher_item SET ?';
+
+  transaction.addQuery(queryVoucherItem, [req.body.voucher_item]);
+
+  transaction.execute()
+  .then(function (rows) {
+    res.status(201).json({ id: req.body.voucher.uuid });
+  })
+  .catch(next)
+  .done();
+}
+
+/**
+* @function checkForMissing
+* @param {object} The sended object
+* @return {boolean}
+* @description This functin check missing data for voucher and voucher item
+*/
+function checkForMissing(_data) {
+  var voucher = _data.voucher;
+  var voucher_item = _data.voucher_item;
+
+  var validVoucher =
```

Comment:
Can do find a case where the database accepts an invalid voucher?  This code is complex and ideally should be automatically done in the database.  Why does this code exist?

My suggestion: write an integration tests that shows that the database will not reject an invalid voucher.  That will be the best way for us to test it on various machines and determine the cause of the problem.